source/simulator/introspection.cc

```diff
@@ -32,166 +32,157 @@ namespace aspect
 {
   namespace internal
   {
-
     /**
-     * Return pair with @p n_components and a filled ComponentIndices structure.
+     * Return a filled ComponentIndices structure.
      */
     template <int dim>
-    std::pair<unsigned int, typename Introspection<dim>::ComponentIndices>
-    setup_component_indices (const unsigned int n_compositional_fields)
+    typename Introspection<dim>::ComponentIndices
+    setup_component_indices (dealii::FEVariableSystem<dim> &ib)
     {
       typename Introspection<dim>::ComponentIndices ci;
-      unsigned int comp = 0;
-
       for (unsigned int i=0; i<dim; ++i)
-        ci.velocities[i] = comp++;
-      ci.pressure = comp++;
-      ci.temperature = comp++;
+        ci.velocities[i] = ib.variable("velocity").first_component_index+i;
+
+      ci.pressure = ib.variable("pressure").first_component_index;
+      ci.temperature = ib.variable("temperature").first_component_index;
+      unsigned int n_compositional_fields = ib.variable("compositions").n_components();
       for (unsigned int i=0; i<n_compositional_fields; ++i)
-        ci.compositional_fields.push_back(comp++);
-      return std::make_pair(comp, ci);
+        ci.compositional_fields.push_back(ib.variable("compositions").first_component_index+i);
```

Comment:
I could imagine models where we do not need the temperature, in case it's only a mechanical model. Of course, as Timo already mentioned, right now there are a lot of places in the code where we use things like `introspection.component_indices.temperature`, and the plugin would have to replace them. So I am totally fine with suggestion 1, it was just an idea that I had.